src/backend/cuda/kernel/anisotropic_diffusion.hpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <common/dispatch.hpp>
+#include <Param.hpp>
+#include <debug_cuda.hpp>
+#include <math.hpp>
+#include <ops.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+static const int THREADS_X = 32;
+static const int THREADS_Y = 8;
+
+inline __device__
+int clamp(int value, int low, int high)
+{
+    return max(low, min(value, high));
+}
+
+__forceinline__ __device__
+int index(int x, int y, int dim0, int dim1, int stride0, int stride1)
+{
+    return clamp(x, 0, dim0-1)*stride0 + clamp(y, 0, dim1-1)*stride1;
+}
+
+__device__
+float quad(float value)
+{
+    return 1.0/(1.0+value);
+}
+
+template<af_flux_function fftype>
+__device__
+float computeGradientBasedUpdate(float mct, float C,
+                                 float S, float N, float W, float E,
+                                 float SE, float SW, float NE, float NW)
+{
+    float delta = 0;
+
+    float dx, dy, df, db, cx, cxd;
+
+    // centralized derivatives
+    dx = (E-W)*0.5f;
+    dy = (S-N)*0.5f;
+
+    // half-d's and conductance along first dimension
+    df  = E - C;
+    db  = C - W;
+
+    if (fftype==AF_FLUX_EXPONENTIAL) {
+        cx  = expf( (df*df + 0.25f*powf(dy+0.5f*(SE - NE), 2)) * mct );
+        cxd = expf( (db*db + 0.25f*powf(dy+0.5f*(SW - NW), 2)) * mct );
+    } else {
+        cx  = quad( (df*df + 0.25f*powf(dy+0.5f*(SE - NE), 2)) * mct );
+        cxd = quad( (db*db + 0.25f*powf(dy+0.5f*(SW - NW), 2)) * mct );
+    }
+    delta += (cx*df - cxd*db);
+
+    // half-d's and conductance along second dimension
+    df  = S - C;
+    db  = C - N;
+
+    if (fftype==AF_FLUX_EXPONENTIAL) {
+        cx  = expf( (df*df + 0.25f*powf(dx+0.5f*(SE - SW), 2)) * mct );
+        cxd = expf( (db*db + 0.25f*powf(dx+0.5f*(NE - NW), 2)) * mct );
+    } else {
+        cx  = quad( (df*df + 0.25f*powf(dx+0.5f*(SE - SW), 2)) * mct );
+        cxd = quad( (db*db + 0.25f*powf(dx+0.5f*(NE - NW), 2)) * mct );
+    }
+    delta += (cx*df - cxd*db);
+
+    return delta;
+}
+
+template<af_flux_function fftype>
```

Comment:
Same as above. Can you make sure you are going to see an increase in perf before making this a template argument. I want to reduce the number of instantiations where ever possible.